datafusion/sqllogictest/test_files/coalesce.slt

```diff
@@ -209,28 +209,20 @@ select
 [3, 4] List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })
 
 # TODO: after switch signature of array to the same with coalesce, this query should be fixed
-query ?T
+query error
 select
   coalesce(array[1, 2], array[arrow_cast(3, 'Int32'), arrow_cast(4, 'Int32')]),
   arrow_typeof(coalesce(array[1, 2], array[arrow_cast(3, 'Int32'), arrow_cast(4, 'Int32')]));
-----
-[1, 2] List(Field { name: "item", data_type: Int64, nullable: true, dict_id: 0, dict_is_ordered: false, metadata: {} })
 
 statement ok
 create table test1 as values (arrow_cast('foo', 'Dictionary(Int32, Utf8)')), (null);
 
 # Dictionary and String are not coercible
-query ?
+query error
```

Comment:
Can we please keep this functionality?  (we use dictionaries heavily in IOx -- the user doesn't really know or hopefully have to care if a column is `Dictionary` or `String`)

I don't see why we can't keep coercing string dictionaries to strings and visa versa

Reply:
I think dictionary type (composite type including struct) is not implicitly cast to other numeric type.

For example,
```
postgres=# select array[row(1, 'a'), 'b'];
ERROR:  input of anonymous composite types is not implemented
LINE 1: select array[row(1, 'a'), 'b'];
                                  ^
postgres=# select array[row(1, 'a')::TEXT, 'b'];
    array    
-------------
 {"(1,a)",b}
(1 row)
```

But, given it is already used heavily and implicitly casting is also not a big issue, I think I also prefer to keep it unchanged.

Reply:
I recently met an similar issue in moving aggregate function where they have their owned coercion rule, specifically `Sum` and `Avg`.

https://github.com/apache/datafusion/blob/96487ea0cbb7901a1e4aa18fdf6deb8961319fea/datafusion/expr/src/type_coercion/aggregates.rs#L108-L149

I think we need to have function-specific coercion. Introduce TypeSignature::coercion, and it calls `fn coerce_types` for udf, udaf function in `coerce_arguments_for_signature`. We can then have different coercion rule for coalesce, make_array, sum, avg. This not only solve the issue here and also for UDAF.

What do you think? @alamb

Reply:
Filed https://github.com/apache/datafusion/issues/10423 to track